istio/discovery.go

```diff
@@ -0,0 +1,565 @@
+		return ""
+	}
+
+	globalConfig, ok := parsedConfig["global"]
+	if !ok {
+		// This does not return an error, because it's probably valid that the configmap does not have the "values.global" key.
+		// So, tell that the network wasn't found by returning blank values
+		log.Debugf("Cannot resolve the network ID of the cluster where Kiali is hosted: no global configuration found for the sidecar injector.")
+		return ""
+	}
+
+	typedGlobalConfig, ok := globalConfig.(map[string]interface{})
+	if !ok {
+		log.Debug("cannot parse the config map of the Istio sidecar injector")
+		return ""
+	}
+
+	networkConfig, ok := typedGlobalConfig["network"]
+	if !ok {
+		// This does not return an error, because it's valid that the configmap does not have the "values.global.network" key, which most
+		// likely means that Istio is not setup for multi-clustering.
+		// So, tell that the network wasn't found by returning blank values
+		log.Debugf("Cannot resolve the network ID of the cluster where Kiali is hosted: multi-cluster is probably turned off.")
+		return ""
+	}
+
+	typedNetworkConfig, ok := networkConfig.(string)
+	if !ok {
+		// It's probably invalid that the network id is not a string
+		return ""
+	}
+
+	return typedNetworkConfig
+}
+
+// resolveNetwork tries to resolve the NETWORK_ID (as known by the Control Plane) of the
+// cluster that can be accessed using the provided kubeconfig data.
+// Also, it's assumed that the control plane on the remote cluster is hosted in the same
+// namespace as in Kiali's Home cluster.
+//
+// No errors are returned because we don't want to block processing of other clusters if
+// one fails. So, errors are only logged to let processing continue.
+func (in *Discovery) resolveNetwork(clusterName string) string {
+	kubeCache, err := in.kialiCache.GetKubeCache(clusterName)
+	if err != nil {
+		log.Warningf("Cannot resolve the network ID of the cluster [%s]: cannot get the kube cache: %v", clusterName, err)
+		return ""
+	}
+
+	if network := in.getNetworkFromSidecarInejctorConfigMap(kubeCache); network != "" {
+		return network
+	}
+
+	// Network id wasn't found in the config. Try to find it on the istio namespace.
+
+	// Let's assume that the istio namespace has the same name on all clusters in the mesh.
+	istioNamespace, err := in.kialiSAClients[clusterName].GetNamespace(in.conf.IstioNamespace)
+	if err != nil {
+		log.Warningf("Cannot describe the [%s] namespace on cluster [%s]: %v", in.conf.IstioNamespace, clusterName, err)
+		return ""
+	}
+
+	// For Kiali's control plane, we used the istio sidecar injector config map to fetch the network ID. This
+	// approach is probably more accurate, because that's what is injected along the sidecar. However,
+	// in remote clusters, we don't have privileges to query config maps, so it's not possible to fetch
+	// the sidecar injector config map. However, Istio docs say that the Istio namespace must be labeled with
```

Comment:
This also helps in the case of ambient-only meshes where there are no sidecars/sidecar config maps (unless there is always a sidecar config map regardless of ambient mode; I don't know).